src/push.rs

```diff
@@ -0,0 +1,197 @@
+            Config{
+                max_idle: HYPER_MAX_IDLE,
+            }
+        );
+}
+
+/// `push_metrics` pushes all gathered metrics to the Pushgateway specified by
+/// url, using the provided job name and the (optional) further grouping labels
+/// (the grouping map may be nil). See the Pushgateway documentation for
+/// detailed implications of the job and other grouping labels. Neither the job
+/// name nor any grouping label value may contain a "/". The metrics pushed must
+/// not contain a job label of their own nor any of the grouping labels.
+///
+/// You can use just host:port or ip:port as url, in which case 'http://' is
+/// added automatically. You can also include the schema in the URL. However, do
+/// not include the '/metrics/jobs/...' part.
+///
+/// Note that all previously pushed metrics with the same job and other grouping
+/// labels will be replaced with the metrics pushed by this call. (It uses HTTP
+/// method 'PUT' to push to the Pushgateway.)
+pub fn push_metrics(job: &str,
+                    grouping: HashMap<String, String>,
+                    url: &str,
+                    mfs: Vec<proto::MetricFamily>)
+                    -> Result<()> {
+    push(job, grouping, url, mfs, "PUT")
+}
+
+/// `push_add_metrics` works like `push_metrics`, but only previously pushed
+/// metrics with the same name (and the same job and other grouping labels) will
+/// be replaced. (It uses HTTP method 'POST' to push to the Pushgateway.)
+pub fn push_add_metrics(job: &str,
+                        grouping: HashMap<String, String>,
+                        url: &str,
+                        mfs: Vec<proto::MetricFamily>)
+                        -> Result<()> {
+    push(job, grouping, url, mfs, "POST")
+}
+
+fn push(job: &str,
+        grouping: HashMap<String, String>,
+        url: &str,
+        mfs: Vec<proto::MetricFamily>,
+        method: &str)
+        -> Result<()> {
+
+    let mut push_url = if url.contains("://") {
+        url.to_owned()
+    } else {
+        format!("http://{}", url)
+    };
+
+    if push_url.ends_with('/') {
+        push_url.pop();
+    }
+
+    let mut url_components = Vec::new();
+    if job.contains('/') {
+        return Err(Error::Msg(format!("job contains '/': {}", job)));
+    }
+
+    // TODO: escape job
+    url_components.push(job.to_owned());
+
+    for (ln, lv) in grouping {
+        // TODO: check label name
+        if lv.contains('/') {
+            return Err(Error::Msg(format!("value of grouping label {} contains '/': {}", ln, lv)));
+        }
+        url_components.push(ln.to_owned());
+        url_components.push(lv.to_owned());
+    }
+
+    push_url = format!("{}/metrics/job/{}", push_url, url_components.join("/"));
+
+    let encoder = TextEncoder::new();
+    let mut buf = Vec::new();
+    try!(encoder.encode(&mfs, &mut buf));
+
+    let request = HTTP_CLIENT.request(Method::from_str(method).unwrap(), &push_url)
```

Comment:
No need to check for pre-existing grouping labels?
https://github.com/prometheus/client_golang/blob/master/prometheus/push/push.go#L103